src/realm/collection.hpp

```diff
@@ -122,231 +181,230 @@ struct AverageHelper<T, std::void_t<ColumnSumType<T>>> {
     }
 };
 
-class CollectionBase : public ArrayParent {
+/// Convenience base class for collections, which implements most of the
+/// relevant interfaces for a collection that is bound to an object accessor and
+/// representable as a BPlusTree<T>.
+template <class Interface>
+class CollectionBaseImpl : public Interface, protected ArrayParent {
 public:
-    virtual ~CollectionBase();
-    CollectionBase(const CollectionBase&) = default;
+    static_assert(std::is_base_of_v<CollectionBase, Interface>);
 
-    /*
-     * Operations that makes sense without knowing the specific type
-     * can be made virtual.
-     */
-    virtual size_t size() const = 0;
-    virtual bool is_null(size_t ndx) const = 0;
-    virtual Mixed get_any(size_t ndx) const = 0;
-    virtual void clear() = 0;
+    // Overriding members of CollectionBase:
+    ColKey get_col_key() const noexcept final
+    {
+        return m_col_key;
+    }
 
-    virtual Mixed min(size_t* return_ndx = nullptr) const = 0;
-    virtual Mixed max(size_t* return_ndx = nullptr) const = 0;
-    virtual Mixed sum(size_t* return_cnt = nullptr) const = 0;
-    virtual Mixed avg(size_t* return_cnt = nullptr) const = 0;
+    TableRef get_target_table() const final
+    {
+        return m_obj.get_target_table(m_col_key);
+    }
 
+    const Obj& get_obj() const noexcept final
+    {
+        return m_obj;
+    }
 
-    std::unique_ptr<CollectionBase> clone() const
+    ObjKey get_key() const noexcept final
     {
-        return m_obj.get_collection_ptr(m_col_key);
+        return m_obj.get_key();
     }
 
-    TableRef get_target_table() const
+    bool is_attached() const noexcept final
     {
-        return m_obj.get_target_table(m_col_key);
+        return m_obj.is_valid();
     }
 
-    // Modifies a vector of indices so that they refer to values sorted according
-    // to the specified sort order
-    virtual void sort(std::vector<size_t>& indices, bool ascending = true) const = 0;
-    // Modifies a vector of indices so that they refer to distinct values.
-    // If 'sort_order' is supplied, the indices will refer to values in sort order,
-    // otherwise the indices will be in original order.
-    virtual void distinct(std::vector<size_t>& indices, util::Optional<bool> sort_order = util::none) const = 0;
+    bool has_changed() const noexcept final
+    {
+        update_if_needed();
```

Comment:
Yeah, it will just terminate. What's more worrying is that it's very confusing, because the method name does not suggest that any internal state is mutated.

Reply:
I've removed the `noexcept`.